README.md
# Image Proxy

[![NPM version](https://badge.fury.io/js/image-proxy.svg)](https://badge.fury.io/js/image-proxy)
[![Build Status](https://secure.travis-ci.org/jpmckinney/image-proxy.png)](https://travis-ci.org/jpmckinney/image-proxy)
[![Dependency Status](https://david-dm.org/jpmckinney/image-proxy.svg)](https://david-dm.org/jpmckinney/image-proxy)
[![Coverage Status](https://coveralls.io/repos/jpmckinney/image-proxy/badge.png)](https://coveralls.io/r/jpmckinney/image-proxy)

A simple Express app for proxying and manipulating images, specifically headshots.

The code is just over 100 lines, making it easy to tailor to your needs.

## Getting Started

First, [install the dependencies](https://github.com/aheckmann/gm#getting-started) for the `gm` package.

    npm install
    npm start
    curl -I http://localhost:5000/http%3A%2F%2F1.gravatar.com%2Favatar%2F60f641dfbb4215f1f6d6c059eebf1848/240/80.jpg

The URL structure is `/:url/:width/:height.:extension?`. The `:url` parameter must be escaped/encoded. If the remote image's width or height is greater than the given `:width` or `:height`, it will be resized, maintaining aspect ratio, and cropped. If smaller, it will be padded with white pixels. If an optional `:extension` parameter is provided, the image will be transcoded to the corresponding file format. The equivalent ImageMagick command for the example URL above is:

    convert input.jpg -thumbnail 240x80^> -gravity center -extent 240x80 output.jpg

The `Cache-Control` header sets a `max-age` of one year.

## Features

Image proxy:

* Supports HTTP and HTTPS
* Follows 301 and 302 redirects
* Sets a maximum timeout for the remote server
* Handles complex MIME types like `image/jpeg; charset=utf-8`
* Optional whitelisting using regular expressions

Image manipulation:

* Accepts a custom width and height, up to 1000x1000
* Resizes, centers and crops the image
* Optionally transcodes to another file format

HTTP server:

* No query string parameters (preferred by CloudFront)
* Adds a Cache-Control header

If you need more features, see [node-imageable](https://github.com/sdepold/node-imageable) and [node-imageable-server](https://github.com/dawanda/node-imageable-server).

### Environment variables

* `DELAY`: The timeout delay in milliseconds, after which the proxy will respond with a HTTP 504 Gateway Timeout server error. Default: `5000`
* `WHITELIST`: A comma-separated list of domains to whitelist, e.g. `.gov,facebook.com`, which will be transformed into the regular expressions `/\.gov$/` and `/facebook\.com$/`.
* `PORT`: If running the server, changes the port on which it listens. Default: `5000`

## Deployment

### Heroku

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

    git clone https://github.com/jpmckinney/image-proxy.git
    heroku apps:create
    heroku config:set NODE_ENV=production
    git push heroku master
    heroku apps:open

### AWS CloudFront

Create a distribution and set the "Origin Domain Name" to the domain name of your Heroku app.

## Testing

    npm test

## Acknowledgements

This project is inspired by [node-connect-image-proxy](https://github.com/mysociety/node-connect-image-proxy).

Copyright (c) 2013 James McKinney, released under the MIT license


CHANGELOG.md
# Changelog

## Unreleased

* Fix transcoding to another file format [#21](https://github.com/jpmckinney/image-proxy/pull/22), thanks @equivalentideas

## 0.0.7 (2015-09-22)

* Set a `Accept: */*` header [#15](https://github.com/jpmckinney/image-proxy/pull/15)

## 0.0.6 (2015-04-16)

* [Set colorspace to RGB before resizing](http://www.imagemagick.org/Usage/resize/#resize_colorspace), thanks @lizconlan

## 0.0.5 (2015-03-30)

* Add transcoding support

## 0.0.4 (2015-03-11)

* Set a `User-Agent` header

## 0.0.3 (2014-01-16)

* Handles relative redirects
* Fix bug in handling timeouts [#9](https://github.com/jpmckinney/image-proxy/issues/9)
* Don't use socket pooling to avoid `maxSockets` limit [#10](https://github.com/jpmckinney/image-proxy/issues/10)

## 0.0.2 (2014-10-06)

* Fix parsing of `WHITELIST` environment variable [#7](https://github.com/jpmckinney/image-proxy/pull/7)
* Upgrade to Express >=3
* Rename and move files
* Add tests

## 0.0.1 (Unpublished)

First release


lib/image-proxy.js
// @see https://devcenter.heroku.com/articles/nodejs#write-your-app

var express = require('express')
  , fs      = require('fs') // node
  , gm      = require('gm')
  , http    = require('http') // node
  , https   = require('https') // node
  , mime    = require('mime')
  , url     = require('url') // node
  // @see http://aaronheckmann.posterous.com/graphicsmagick-on-heroku-with-nodejs
  , app = express()
  , imageMagick = gm.subClass({imageMagick: true})
  , whitelist = process.env.WHITELIST || [] // [/\.gov$/, /google\.com$/]
  , delay = parseInt(process.env.DELAY) || 5000
  , mimeTypes = [
    'image/gif',
    'image/jpeg',
    'image/png',
    // Common typos
    'image/jpg',
  ];

module.exports = function () {

  app.get('/:url/:width/:height.:extension?', function (req, res, next) {
    var width = req.params.width
      , height = req.params.height
      , extension = req.params.extension
      , retrieve = function (remote) {
        // @see http://nodejs.org/api/url.html#url_url
        var options = url.parse(remote);
        // @see https://github.com/substack/hyperquest
        options.agent = false;
        if (options.protocol !== 'http:' && options.protocol !== 'https:') {
          return res.status(404).send('Expected URI scheme to be HTTP or HTTPS');
        }
        if (!options.hostname) {
          return res.status(404).send('Expected URI host to be non-empty');
        }
        options.headers = {'User-Agent': 'image-proxy/0.0.7', 'Accept': '*/*'};

        var agent = options.protocol === 'http:' ? http : https
          , timeout = false
          // @see http://nodejs.org/api/http.html#http_http_get_options_callback
          , request = agent.get(options, function (response) {
            if (timeout) {
              // Status code 504 already sent.
              return;
            }

            // @see http://nodejs.org/api/http.html#http_response_statuscode
            if ((response.statusCode === 301 || response.statusCode === 302) && response.headers['location']) {
              var redirect = url.parse(response.headers['location']);
              // @see https://tools.ietf.org/html/rfc7231#section-7.1.2
              if (!redirect.protocol) {
                redirect.protocol = options.protocol;
              }
              if (!redirect.hostname) {
                redirect.hostname = options.hostname;
              }
              if (!redirect.port) {
                redirect.port = options.port;
              }
              if (!redirect.hash) {
                redirect.hash = options.hash;
              }
              return retrieve(url.format(redirect));
            }

            // The image must return status code 200.
            if (response.statusCode !== 200) {
              return res.status(404).send('Expected response code 200, got ' + response.statusCode);
            }

            // The image must be a valid content type.
            // @see http://nodejs.org/api/http.html#http_request_headers
            var mimeType;
            if (extension) {
              mimeType = mime.lookup(extension);
            }
            else {
              mimeType = (response.headers['content-type'] || '').replace(/;.*/, '');
              extension = mime.extension(mimeType);
            }
            if (mimeTypes.indexOf(mimeType) === -1) {
              return res.status(404).send('Expected content type ' + mimeTypes.join(', ') + ', got ' + mimeType);
            }

            // @see https://github.com/aheckmann/gm#constructor
            imageMagick(response, 'image.' + extension)
            .colorspace('RGB')
            // @see http://www.imagemagick.org/Usage/thumbnails/#cut
            .resize(width, height + '^>')
            .gravity('Center') // faces are most often near the center
            .extent(width, height)
            .stream(extension, function (err, stdout, stderr) {
              if (err) return next(err);
              // Log errors in production.
              stderr.pipe(process.stderr);
              // @see http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html
              res.writeHead(200, {
                'Content-Type': mimeType,
                'Cache-Control': 'max-age=31536000, public', // 1 year
              });
              stdout.pipe(res);
            });
          }).on('error', next);

        // Timeout after five seconds. Better luck next time.
        request.setTimeout(delay, function () {
          timeout = true; // if we abort, we'll get a "socket hang up" error
          return res.status(504).send();
        });
      };

    // Validate parameters.
    if (whitelist.length) {
      var parts = url.parse(req.params.url);
      if (parts.hostname) {
        var any = false, _i, _len;
        if (typeof whitelist === 'string') {
          whitelist = whitelist.split(',');
        }
        for (_i = 0, _len = whitelist.length; _i < _len; _i++) {
          if (typeof whitelist[_i] === 'string') {
            // Escape periods and add anchor.
            whitelist[_i] = new RegExp(whitelist[_i].replace('.', '\\.') + '$')
          }
          if (whitelist[_i].test(parts.hostname)) {
            any = true;
            break;
          }
        }
        if (!any) { // if none
          return res.status(404).send('Expected URI host to be whitelisted');
        }
      }
    }
    if (isNaN(parseInt(width))) {
      return res.status(404).send('Expected width to be an integer');
    }
    if (parseInt(width) > 1000) {
      return res.status(404).send('Expected width to be less than or equal to 1000');
    }
    if (isNaN(parseInt(height))) {
      return res.status(404).send('Expected height to be an integer');
    }
    if (parseInt(height) > 1000) {
      return res.status(404).send('Expected height to be less than or equal to 1000');
    }

    retrieve(req.params.url);
  });

  return app;
};
